<a href="https://colab.research.google.com/github/muid-uddin-chowdhury/ETE_456/blob/main/Lab_05_Classification_1608030.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


ETE-456: Classification Project on Stroke Prediction Dataset

    Objective:

        Apply various classification algorithms on a real world dataset.



Stroke Prediction Dataset

Context: According to the World Health Organization (WHO) stroke is the 2nd leading cause of death globally, responsible for approximately 11% of total deaths. This dataset is used to predict whether a patient is likely to get stroke based on the input parameters like gender, age, various diseases, and smoking status. Each row in the data provides relavant information about the patient.
**Attribute Information**

    1)id: unique identifier
    2)gender: "Male", "Female" or "Other"
    3)age: age of the patient
    4)hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension
    5)heart_disease: 0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
    6)ever_married: "No" or "Yes"
    7)work_type: "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
    8)Residence_type: "Rural" or "Urban"
    9)avg_glucose_level: average glucose level in blood
    10)bmi: body mass index
    11)smoking_status: "formerly smoked", "never smoked", "smokes" or "Unknown"*
    12)stroke: 1 if the patient had a stroke or 0 if not *Note: "Unknown" in smoking_status means that the information is unavailable for this patient


In [ ]:
import warnings
warnings.filterwarnings("ignore")

Import Libaries

In [ ]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt

Dataset

In [ ]:
!wget -O stroke-data.csv https://www.dropbox.com/s/zgburk3yces5tee/healthcare-dataset-stroke-data.csv?dl=0

--2021-12-21 04:23:29--  https://www.dropbox.com/s/zgburk3yces5tee/healthcare-dataset-stroke-data.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.71.18, 2620:100:6027:18::a27d:4812
Connecting to www.dropbox.com (www.dropbox.com)|162.125.71.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/zgburk3yces5tee/healthcare-dataset-stroke-data.csv [following]
--2021-12-21 04:23:29--  https://www.dropbox.com/s/raw/zgburk3yces5tee/healthcare-dataset-stroke-data.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc9c7fabcc40420e1be56357fc62.dl.dropboxusercontent.com/cd/0/inline/BcMHprK7O9tNsXy8dbf8lNew3_Fl4Ka3prVx9_QzdK-Vae2_0BR2KIF6gwtVoca5QK13s6BmIchmaEK4gn6dqsOm11vp-6n2Z446wAO7T44Xoe7SJ_shYD7tmEfd5Xp0jzr3Oow6T3VRJHtY1cz7W0fD/file# [following]
--2021-12-21 04:23:30--  https://uc9c7fabcc40420e1be56357fc62.dl.dropboxusercontent.com/cd/0/inline/BcMHprK7O9tNsXy8

In [ ]:
dataset = pd.read_csv('stroke-data.csv')
dataset

id  gender   age  ...   bmi   smoking_status stroke
0      9046    Male  67.0  ...  36.6  formerly smoked      1
1     51676  Female  61.0  ...   NaN     never smoked      1
2     31112    Male  80.0  ...  32.5     never smoked      1
3     60182  Female  49.0  ...  34.4           smokes      1
4      1665  Female  79.0  ...  24.0     never smoked      1
...     ...     ...   ...  ...   ...              ...    ...
5105  18234  Female  80.0  ...   NaN     never smoked      0
5106  44873  Female  81.0  ...  40.0     never smoked      0
5107  19723  Female  35.0  ...  30.6     never smoked      0
5108  37544    Male  51.0  ...  25.6  formerly smoked      0
5109  44679  Female  44.0  ...  26.2          Unknown      0

[5110 rows x 12 columns]

In [ ]:
features = dataset[['age',	'hypertension',	'heart_disease',	'bmi','Residence_type','avg_glucose_level',	'smoking_status']]  # for independent
target = dataset[['stroke']]

In [ ]:
features

age  hypertension  ...  avg_glucose_level   smoking_status
0     67.0             0  ...             228.69  formerly smoked
1     61.0             0  ...             202.21     never smoked
2     80.0             0  ...             105.92     never smoked
3     49.0             0  ...             171.23           smokes
4     79.0             1  ...             174.12     never smoked
...    ...           ...  ...                ...              ...
5105  80.0             1  ...              83.75     never smoked
5106  81.0             0  ...             125.20     never smoked
5107  35.0             0  ...              82.99     never smoked
5108  51.0             0  ...             166.29  formerly smoked
5109  44.0             0  ...              85.28          Unknown

[5110 rows x 7 columns]

In [ ]:
target

stroke
0          1
1          1
2          1
3          1
4          1
...      ...
5105       0
5106       0
5107       0
5108       0
5109       0

[5110 rows x 1 columns]

Mising Values Implementations

In [ ]:
from sklearn.impute import SimpleImputer

In [ ]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean")
imputer = imputer.fit(features[['age','hypertension',	'heart_disease','bmi','avg_glucose_level']])

In [ ]:
features[['age',	'hypertension',	'heart_disease','bmi','avg_glucose_level']]= imputer.transform(features[['age',	'hypertension',	'heart_disease','bmi','avg_glucose_level']])

In [ ]:
features

age  hypertension  ...  avg_glucose_level   smoking_status
0     67.0           0.0  ...             228.69  formerly smoked
1     61.0           0.0  ...             202.21     never smoked
2     80.0           0.0  ...             105.92     never smoked
3     49.0           0.0  ...             171.23           smokes
4     79.0           1.0  ...             174.12     never smoked
...    ...           ...  ...                ...              ...
5105  80.0           1.0  ...              83.75     never smoked
5106  81.0           0.0  ...             125.20     never smoked
5107  35.0           0.0  ...              82.99     never smoked
5108  51.0           0.0  ...             166.29  formerly smoked
5109  44.0           0.0  ...              85.28          Unknown

[5110 rows x 7 columns]

In [ ]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean")
imputer = imputer.fit(target[['stroke']])

In [ ]:
target[['stroke']]= imputer.transform(target[['stroke']])

In [ ]:
target

stroke
0        1.0
1        1.0
2        1.0
3        1.0
4        1.0
...      ...
5105     0.0
5106     0.0
5107     0.0
5108     0.0
5109     0.0

[5110 rows x 1 columns]

Encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [ ]:
labelencoder_f = LabelEncoder()

features['Residence_type'] = labelencoder_f.fit_transform(features['Residence_type'])

In [ ]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['smoking_status']]))

In [ ]:
encoded_labels.columns = encoder.get_feature_names(['smoking_status'])
dataset= pd.concat([features, encoded_labels ], axis=1)

In [ ]:
dataset

age  hypertension  ...  smoking_status_never smoked  smoking_status_smokes
0     67.0           0.0  ...                          0.0                    0.0
1     61.0           0.0  ...                          1.0                    0.0
2     80.0           0.0  ...                          1.0                    0.0
3     49.0           0.0  ...                          0.0                    1.0
4     79.0           1.0  ...                          1.0                    0.0
...    ...           ...  ...                          ...                    ...
5105  80.0           1.0  ...                          1.0                    0.0
5106  81.0           0.0  ...                          1.0                    0.0
5107  35.0           0.0  ...                          1.0                    0.0
5108  51.0           0.0  ...                          0.0                    0.0
5109  44.0           0.0  ...                          0.0                    0.0

[5110 rows x 11 columns]

In [ ]:
dataset.columns

Index(['age', 'hypertension', 'heart_disease', 'bmi', 'Residence_type',
       'avg_glucose_level', 'smoking_status', 'smoking_status_Unknown',
       'smoking_status_formerly smoked', 'smoking_status_never smoked',
       'smoking_status_smokes'],
      dtype='object')

In [ ]:
new_features = dataset[['age', 'hypertension', 'heart_disease', 'bmi',
       'avg_glucose_level','Residence_type',
       'smoking_status_Unknown',
       'smoking_status_formerly smoked', 'smoking_status_never smoked',
       'smoking_status_smokes']]

Splitting Dataset

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size=0.3,random_state=0)

In [ ]:
print(X_train.shape)
print(X_test.shape)

(3577, 10)
(1533, 10)


In [ ]:
X_test

age  hypertension  ...  smoking_status_never smoked  smoking_status_smokes
42    82.0           0.0  ...                          0.0                    1.0
380    4.0           0.0  ...                          0.0                    0.0
3524  58.0           0.0  ...                          1.0                    0.0
352   20.0           0.0  ...                          1.0                    0.0
4927  10.0           0.0  ...                          0.0                    0.0
...    ...           ...  ...                          ...                    ...
929   75.0           0.0  ...                          0.0                    0.0
2629  45.0           0.0  ...                          0.0                    0.0
663   38.0           0.0  ...                          1.0                    0.0
2303  75.0           0.0  ...                          0.0                    1.0
3122  57.0           1.0  ...                          0.0                    1.0

[1533 rows x 10 columns]

Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()

lr.fit(X_train,y_train)

LogisticRegression()

In [ ]:
y_pred = lr.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score,precision_score,recall_score

In [ ]:
confusion_matrix(y_test,y_pred)

array([[1457,    0],
       [  76,    0]])

In [ ]:
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.00      0.00      0.00        76

    accuracy                           0.95      1533
   macro avg       0.48      0.50      0.49      1533
weighted avg       0.90      0.95      0.93      1533



KNN Classifiers

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

classifier = KNeighborsClassifier(n_neighbors=3, metric = 'minkowski')
classifier.fit(X_train,y_train)

y_pred = classifier.predict(X_test)

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[1431   26]
 [  71    5]]


In [ ]:
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      0.98      0.97      1457
         Yes       0.16      0.07      0.09        76

    accuracy                           0.94      1533
   macro avg       0.56      0.52      0.53      1533
weighted avg       0.91      0.94      0.92      1533



Support Vector Machine (SVM)

In [ ]:
from sklearn.svm import SVC

classifier = SVC(kernel = 'linear', random_state = 42)
classifier.fit(X_train,y_train)

y_pred = classifier.predict(X_test)

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[1457    0]
 [  76    0]]


In [ ]:
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.00      0.00      0.00        76

    accuracy                           0.95      1533
   macro avg       0.48      0.50      0.49      1533
weighted avg       0.90      0.95      0.93      1533



Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train,y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [ ]:
y_pred = classifier.predict(X_test)

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[1395   62]
 [  64   12]]


In [ ]:
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.96      0.96      0.96      1457
         Yes       0.16      0.16      0.16        76

    accuracy                           0.92      1533
   macro avg       0.56      0.56      0.56      1533
weighted avg       0.92      0.92      0.92      1533



Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(n_estimators=32, criterion ='entropy', random_state = 40)
classifier.fit(X_train,y_train)

RandomForestClassifier(criterion='entropy', n_estimators=32, random_state=40)

In [ ]:
y_pred = classifier.predict(X_test)

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[1456    1]
 [  74    2]]


In [ ]:
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.67      0.03      0.05        76

    accuracy                           0.95      1533
   macro avg       0.81      0.51      0.51      1533
weighted avg       0.94      0.95      0.93      1533




Result Analysis **bold text**

Here, 5 different algorithms are used and 4 evaluation metrics are used to find out the accuracy. Tha detailed data table is given below. 

**Logistic Regression**

    Weighted Average Precision : 90%
    Weighted Average Recall : 95%
    Weighted Average F1-Score : 93%
    Accuracy : 95%

**KNN Classifier**

    Weighted Average Precision : 91%
    Weighted Average Recall : 94%
    Weighted Average F1-Score : 92%
    Accuracy : 94%

**Support Vector Machine**

    Weighted Average Precision : 90%
    Weighted Average Recall : 95%
    Weighted Average F1-Score : 93%
    Accuracy : 95%

**Decision Tree Classification**

    Weighted Average Precision : 92%
    Weighted Average Recall : 92%
    Weighted Average F1-Score : 92%
    Accuracy : 92%

**Random Forest Classification**

    Weighted Average Precision : 94%
    Weighted Average Recall : 95%
    Weighted Average F1-Score : 93%
    Accuracy : 95%



**Disscussion**

here, for a given dataset named "stroke-data.csv", a machine learning project is ran. Here, like any other machine learning projects, some formalities has to be done as implementing the missing values, label encoding and OneHotEncoding to encode every possible data in binary to make calculations easy for the machine etc. And the splitting of dataset is a random fact that depends on the nature of a data of a dataset. Finally, different algorithms in different evaluation metrics is used and among them, Random Forest Classification can give the best result for this project.